Do not swallow PicklingError

Nielius opened the pull request:
Currently, when caching fails because the function output can't be pickled, there is no warning, but the output will not be cached. This surprises users, who expect their output to be cached.

This is related to https://github.com/joblib/joblib/issues/952 and https://github.com/joblib/joblib/issues/896 . It does not solve the underlying race condition in https://github.com/joblib/joblib/issues/896, but it does implement the suggestion made by @tomMoral [in this comment](https://github.com/joblib/joblib/issues/952#issuecomment-546370327): it just raises an error when the caching fails due to pickling.

tomMoral commented on joblib/_store_backends.py:
```diff
@@ -189,6 +190,8 @@ def write_func(to_write, dest_filename):
                                       compress=self.compress)
 
             self._concurrency_safe_write(item, filename, write_func)
+        except PicklingError:
```
This `try/except` block should be in the `write_func` function.

Nielius replied:
Done

tomMoral reviewed (commented):
Hello,

Thanks a lot for this PR :)
It looks simple enough but  this change could break existing code, so we should do the change in two steps:
- First raise a warning (instead of swallowing the error).
- Then in a later version change to raise the `PicklingError`.

Could you change this to raise a warning instead? Also could you please move the try/except in the `write_func`?

This might be worth it to add a test, to make sure the error is correctly raised.

Nielius commented:
Hi,

Thanks for the quick feedback @tomMoral !

I've addressed your feedback by:

- moving the try/except block
- changing to a warning instead of raising an exception
- adding a unit test to test that a warning is being logged

In addition, I've also changed the catch-all except block to log a warning. When I was writing the test, I noticed how frustrating it was that it literally swallows every kind of error. I think the `warnings.warn( ...)` is better than just doing nothing --- do you agree?

tomMoral commented on joblib/_store_backends.py:
```diff
@@ -185,12 +186,24 @@ def dump_item(self, path, item, verbose=1):
+                            'failed to pickle output. '
+                            'In future versions of joblib '
+                            'this will raise an exception.'
+                            'Exception: %s.' % e,
```
We try to avoid the old formating synthax in the new code base, so please use:
```suggestion
                            f'Exception: {e}.'
```

ogrisel replied:
Also, could you please rewrap the string to better use the available <80 char limit of PEP8?

Some people complain than 80 is not enough, others seem reluctant to use more than 60 apparently ;) 80 is a good middle ground.

tomMoral commented on joblib/_store_backends.py:
```diff
@@ -185,12 +186,24 @@ def dump_item(self, path, item, verbose=1):
             self._concurrency_safe_write(item, filename, write_func)
-        except:  # noqa: E722
-            " Race condition in the creation of the directory "
+        except Exception as e:  # noqa: E722
+            warnings.warn(
+                'Unable to cache to disk. '
+                'Possibly a race condition in the creation of the directory. '
+                'Exception: %s' % e)
```
As we are changing a bit this, I would be in favor of sanitizing this, by putting a `try/except` for `OSError` in `create_location`, and having this raise a deprecation warning saying that any other error will be raised in future version of `joblib`. WDYT?


Small nitpick on style.
```suggestion
        except Exception as e:  # noqa: E722
            warnings.warn(
                'Unable to cache to disk. '
                'Possibly a race condition in the creation of the directory. '
                f'Exception: {e}'
            )
```

tomMoral replied:
So following the conversation with @ogrisel , I would not do a deprecation warning but just a warning for the error.
Maybe create a specfic `CacheWarning` that will be easy to filter out?

tomMoral reviewed (commented):
Thanks for adding the tests!!

LGTM, just a small question for putting a dedicated `try/except` for concurrent folder and raising others in the future?

ogrisel commented:
We could even have a constructor parameter on the `Memory` object: `error="ignore"`, `error="warn"`, or `error="raise"`.

I would be in favor of using `error="warn"` by default but users who prefer other options are free to switch easily.

tomMoral commented:
This feels a bit YAGNI to me. There are already many parameters that users don't understand, I would avoid adding some if possible.
The rational of swallowing error linked to dumping was that caching should not create extra problems due to OS (memory/ disk space/collision). So thinking back, I would say that raising a warning for any error except pickling one is the right behavior.
That way, users can implement the three behaviors simply with warning filters? WDYT?

ogrisel commented:
Alright then.

jeremiedbb commented on joblib/_store_backends.py:
```diff
@@ -185,12 +191,24 @@ def dump_item(self, path, item, verbose=1):
+                    try:
+                        numpy_pickle.dump(to_write, f, compress=self.compress)
+                    except PicklingError as e:
+                        #TODO(1.5) turn into error
```
```suggestion
                        # TODO(1.5) turn into error
```

jeremiedbb reviewed (approved):
LGTM.

@tomMoral, @ogrisel, I applied the requested modifications regarding formatting and added a new warning for all errors but pickling error (it's available from joblib.memory).

tomMoral reviewed (approved):
LGTM! Thanks @Nielius and @jeremiedbb for finishing this.
